routes/api.js

```diff
@@ -139,42 +133,62 @@ router.post('/kernels', bodyParser.json({ type: 'text/plain' }), function(req, r
     initWsProxy(req.connection.server);
 
     // Forward the user object in the session to the kernel gateway.
-    if(config.get('KG_FORWARD_USER_AUTH') && req.user) {
+    if (config.get('KG_FORWARD_USER_AUTH') && req.user) {
         req.body.env = {
              KERNEL_USER_AUTH: JSON.stringify(req.user)
         };
     }
 
-    // Pass the (modified) request to the kernel gateway.
-    request({
-        url: urljoin(kgUrl, kgBaseUrl, '/api/kernels'),
-        method: 'POST',
-        headers: headers,
-        json: req.body
-    }, function(err, response, body) {
-        if(err) {
-            error('Error proxying kernel creation request:' + err.toString());
-            return res.status(500).end();
-        }
-        // Store the notebook path for use within the WS proxy.
-        var notebookPathHeader = req.headers['x-jupyter-notebook-path'];
-        var sessionId = req.headers['x-jupyter-session-id'];
-        if (!notebookPathHeader || !sessionId) {
-            error('Missing notebook path or session ID headers');
-            return res.status(500).end();
-        }
-        var matches = notebookPathHeader.match(/^\/(?:dashboards(-plain)?)?(.*)$/);
-        if (!matches) {
-            error('Invalid notebook path header');
-            return res.status(500).end();
-        }
-        // Store notebook path for later use
-        sessions[sessionId] = matches[2];
+    // Store the notebook path for use within the WS proxy.
+    var notebookPathHeader = req.headers['x-jupyter-notebook-path'];
+    var sessionId = req.headers['x-jupyter-session-id'];
+    if (!notebookPathHeader || !sessionId) {
+        error('Missing notebook path or session ID headers');
+        return res.status(500).end();
+    }
 
-        // Pass the kernel gateway response back to the client.
-        res.set(response.headers);
-        res.status(response.statusCode).json(body);
-    });
+    var matches = notebookPathHeader.match(/^\/(?:dashboards(-plain)?)?(.*)$/);
+    if (!matches) {
+        error('Invalid notebook path header');
+        return res.status(500).end();
+    } else {
+        var notebookPath = matches[2];
+        // Store notebook path for later use
+        sessions[sessionId] = notebookPath;
+
+        nbstore.get(notebookPath)
+        .then(function success(notebook){
+            //debug(notebook.metadata);
+            if (notebook.metadata.kernelspec.name) {
+                var kernelName = notebook.metadata.kernelspec.name;
+                debug("Notebook kernel name found: " + kernelName);
+                if(kernelName === "apache_toree") {
```

Comment:
I see why. The `FROM` line in the `Dockerfile.kernel` is pointing to a build so old, that it predates the Toree kernel and is still using the ET-Spark kernel. Yikes!

So we definitely don't want the translation from apache_toree to scala here. It's making us compatible with the old ET kernel and not Apache Toree.

I suggest moving to 2d878db5cbff which is the last build to have notebook 4.1 and Toree (https://github.com/jupyter/docker-stacks/wiki/Docker-build-history).